lib/Feature/FeatureModelParser.cpp

```diff
@@ -286,10 +275,364 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+                    xmlParserInputBufferCreateMem(
+                        SxfmConstants::DtdRaw.c_str(),
+                        SxfmConstants::DtdRaw.length(), XML_CHAR_ENCODING_UTF8),
+                    XML_CHAR_ENCODING_UTF8),
+      xmlFreeDtd);
+  xmlCleanupParser();
+  assert(Dtd && "Failed to parse DTD.");
+  return Dtd;
+}
+
+FeatureModelSxfmParser::UniqueXmlDoc FeatureModelSxfmParser::parseDoc() {
+  // Initialize the XML parser
+  std::unique_ptr<xmlParserCtxt, void (*)(xmlParserCtxtPtr)> Ctxt(
+      xmlNewParserCtxt(), xmlFreeParserCtxt);
+  // Parse the given model by libxml2
+  UniqueXmlDoc Doc(xmlCtxtReadMemory(Ctxt.get(), Sxfm.c_str(), Sxfm.length(),
+                                     nullptr, nullptr, XML_PARSE_NOBLANKS),
+                   xmlFreeDoc);
+  xmlCleanupParser();
+
+  // In the following, the document is validated.
+  // Therefore, (1) check whether it could be parsed
+  if (Doc && Ctxt->valid) {
+    // (2) validate the sxfm format by using the dtd (document type definition)
+    // file
+    if (xmlValidateDtd(&Ctxt->vctxt, Doc.get(), createDtd().get())) {
+      // and (3) check the tree-like structure of the embedded feature model
+      // as well as constraints
+      return Doc;
+    }
+    llvm::errs() << "Failed to validate DTD.\n";
+  }
+  llvm::errs() << "Failed to parse / validate XML.\n";
+
+  return UniqueXmlDoc(nullptr, nullptr);
+}
+
+bool FeatureModelSxfmParser::parseVm(xmlNode *Node) {
+  // Parse the name first.
+  {
+    UniqueXmlChar Cnt(xmlGetProp(Node, XmlConstants::NAME), xmlFree);
+    FMB.setVmName(std::string(reinterpret_cast<char *>(Cnt.get())));
+  }
+  // After the feature model tag, some metadata is provided.
+  // Currently, we skip the metadata and continue with parsing the feature tree
+
+  for (xmlNode *H = Node->children; H; H = H->next) {
+    if (H->type == XML_ELEMENT_NODE) {
+      if (!xmlStrcmp(H->name, SxfmConstants::FEATURE_TREE)) {
+        // Check whether a custom indentation is defined in the feature tree and
+        // set it accordingly
+        {
+          UniqueXmlChar Cnt(xmlGetProp(H, SxfmConstants::INDENTATION), xmlFree);
+          if (Cnt) {
+            Indentation = reinterpret_cast<char *>(Cnt.get());
+          }
+        }
+
+        // Parse the feature tree with all its features and relations among
+        // them.
+        if (!parseFeatureTree(xmlNodeGetContent(H))) {
+          return false;
+        }
+
+        // Finally, parse the cross-tree-constraints in the constraints tag.
+      } else if (!xmlStrcmp(H->name, SxfmConstants::CONSTRAINTS)) {
+        if (!parseConstraints(xmlNodeGetContent(H))) {
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+  // Split the lines of the feature tree by new lines
+  {
+    std::stringstream Ss(reinterpret_cast<const char *>(FeatureTree));
+    std::string To;
+    string Name;
```

Comment:
```suggestion
    std::string To;
    std::string Name;
```
It looks a bit strange, using one time `std::string` and the other time just `string` (not sure, why it is even available, as I did not saw any `using` at first glance), but that is also pretty inconsistent throughout the whole file (including my code^^).

Reply:
I have changed that. Now it is consistent; at least for the feature model parser.